### PR TITLE
[libc++] Add MSVC's implementation of `exception_ptr` for Windows

### PR DESCRIPTION
**Pull request by StephanTLavavej**
As requested by @petrhosek (and earlier by @llvm-beanz), Microsoft is happy to contribute our implementation of `exception_ptr` for Windows, under the Apache License v2.0 with LLVM Exception that our STLs share.

I've copied our files to plausible-looking places and updated their line endings and banners, but otherwise have made no attempt to integrate them into libc++. I copied all of our `<exception>` even though you only need `exception_ptr` from it. These are our current sources, corresponding to what's going to ship for production use in VS 2022 17.11.

Petr explained to me that libc++ developers will be able to do the integration work - please feel free to push arbitrary changes to my branch. I expect that significant overhauls will be necessary! (Significant improvements should be possible when you don't have to worry about the ancient evil `/clr:pure` etc.)

Please let me know if you need any more information or code. There's a bit of `shared_ptr` integration, where that machinery grants access to `_Exception_ptr_access` defined in `excptptr.cpp` - if you need the `<memory>` side of things, I can add that. (For the record, libc++ is always welcome to grab sources directly from microsoft/STL under our license, but we're also happy to deliver sources like this.)

The only caveat is that our "VCRuntime" support library is not open-source. If you need sources from VCRuntime, we'll have to figure out whether that's possible. Presumably this shouldn't be necessary, since our open-source microsoft/STL consumes the proprietary VCRuntime "at arm's length" and everything works.

Finally, although I've maintained our STL since 2007 and have worked on almost every part of it, I didn't write `exception_ptr` and its inner workings remain mysterious to me. It was written by a Microsoft dev who left the VC Libraries team in 2013 (and later, the company). If you have questions about what it's doing, I can try to get answers, but sometimes your guess will be as good as ours. :joy_cat: There's a potential performance improvement that could be made, which we're currently unable to do for ABI compatibility, see https://github.com/microsoft/STL/issues/751. I don't see any active bugs in our internal database about `exception_ptr`.

## Commits

* microsoft/STL's `excptptr.cpp`, copied verbatim.
  + From our `main` branch as of 2024-06-10: https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/src/excptptr.cpp
* microsoft/STL's `<exception>`, copied verbatim as `win32_exception_ptr.h`.
  + From our `main` branch as of 2024-06-10: https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/inc/exception
* Convert files to LF line endings.
* Add LLVM banners.
* Update `libcxx/CREDITS.TXT` to mention Microsoft.
  + I am intentionally not crediting myself here, I'm just delivering the pizza, I didn't cook it. :pizza:

### CODE REVIEW

File: libcxx/include/__exception/win32_exception_ptr.h [hunk 1]
```diff
@@ -0,0 +1,430 @@
+//===----------------------------------------------------------------------===//
```
Comment (frederick-vs-ja) [inline libcxx/include/__exception/win32_exception_ptr.h:1]: I think the new header and source files should be added to `libcxx/include/CMakeLists.txt` and `libcxx/src/CMakeLists.txt`.

File: libcxx/include/__exception/win32_exception_ptr.h [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _EXCEPTION_
+#define _EXCEPTION_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <cstdlib>
+#include <type_traits>
```
Comment (frederick-vs-ja) [inline libcxx/include/__exception/win32_exception_ptr.h:20]: Should we include granularized internal headers?

File: libcxx/include/__exception/win32_exception_ptr.h [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+_EXPORT_STD extern "C++" _CXX17_DEPRECATE_UNCAUGHT_EXCEPTION _NODISCARD _CRTIMP2_PURE bool __CLRCALL_PURE_OR_CDECL
+    uncaught_exception() noexcept;
+#endif // _HAS_DEPRECATED_UNCAUGHT_EXCEPTION
+_EXPORT_STD extern "C++" _NODISCARD _CRTIMP2_PURE int __CLRCALL_PURE_OR_CDECL uncaught_exceptions() noexcept;
+
+_STD_END
+
+#if _HAS_EXCEPTIONS
+
+#include <malloc.h> // TRANSITION, VSO-2048380: This is unnecessary, but many projects assume it as of 2024-04-29
```
Comment (frederick-vs-ja) [inline libcxx/include/__exception/win32_exception_ptr.h:41]: Is this also needed for libc++ even though the internal header is newly added?

File: libcxx/src/support/win32/excptptr.cpp [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+
+        void _Delete_this() noexcept override {
+            // intentionally does nothing
+        }
+
+    public:
+        // constexpr, TRANSITION P1064
+        explicit _ExceptionPtr_static() noexcept : _Ref_count_base() {
+            _PopulateCppExceptionRecord(_ExRecord, &_Ex, static_cast<ThrowInfo*>(__GetExceptionInfo(_Ex)));
+        }
```
Comment (frederick-vs-ja) [inline libcxx/src/support/win32/excptptr.cpp:226]: It seems that we can never make this ctor `constexpr` as `_PopulateCppExceptionRecord` needs to perform `reinterpret_cast` and `RtlPcToFileHeader` (even if `__GetExceptionInfo` gets `constexpr`-ized one day).
Should we drop this comment (also in MSVC STL)?

**Comment by ldionne** [comment]
@petrhosek Would you be able to pick this up? I can provide guidance if needed.

**Comment by petrhosek** [comment]
> @petrhosek Would you be able to pick this up? I can provide guidance if needed.

This should be ready for review. I used Gemini to automate the refactoring since the original patch was rather large. I made the changes in three stages and I deliberately kept them separate as three commits so they can be reviewed independently and you can more easily trace the changes.

**Comment by ldionne** [comment]
Seems like the CI isn't happy. Let me know when this is ready to review!

**Comment by petrhosek** [comment]
> Seems like the CI isn't happy. Let me know when this is ready to review!

I fixed all test failures and the CI turned green so this is ready for review.

File: libcxx/include/__exception/exception_ptr.h [hunk 5]
```diff
@@ -62,8 +62,6 @@ _LIBCPP_OVERRIDABLE_FUNC_VIS __cxa_exception* __cxa_init_primary_exception(
 _LIBCPP_BEGIN_UNVERSIONED_NAMESPACE_STD
 _LIBCPP_BEGIN_EXPLICIT_ABI_ANNOTATIONS
 
-#ifndef _LIBCPP_ABI_MICROSOFT
```
Comment (ldionne) [inline libcxx/include/__exception/exception_ptr.h:65]: I really like that this patch gives us a single implementation of the public class.

File: libcxx/include/__exception/exception_ptr.h [hunk 6]
```diff
@@ -180,34 +182,7 @@ _LIBCPP_HIDE_FROM_ABI exception_ptr make_exception_ptr(_Ep) _NOEXCEPT {
 
 #else // defined(_LIBCPP_ABI_MICROSOFT)
 
-class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
```
Comment (ldionne) [inline libcxx/include/__exception/exception_ptr.h:183]: We should release-note this change, since IIUC it's ABI-breaking on Windows (since `exception_ptr` changed layout, right?).

**Review by ldionne — APPROVED** [review]
Minor comments, but LGTM. I didn't review the implementation (`exception_pointer_msvc.ipp`) in detail since that part is coming from MSVC itself.

File: libcxx/include/__exception/exception_ptr.h (hunk 6 above)
Comment (petrhosek) [reply libcxx/include/__exception/exception_ptr.h:183]: Updated release notes.

File: libcxx/src/support/runtime/exception_pointer_msvc.ipp [hunk 7]
```diff
@@ -7,70 +7,453 @@
...
 
 nested_exception::nested_exception() noexcept : __ptr_(current_exception()) {}
 
 nested_exception::~nested_exception() noexcept {}
 
 [[noreturn]] void nested_exception::rethrow_nested() const {
-  if (__ptr_ == nullptr)
+  if (__ptr_ == nullptr) {
     terminate();
+  }
```
Comment (frederick-vs-ja) [inline libcxx/src/support/runtime/exception_pointer_msvc.ipp]: These lines don't need to be touched. I _guess_ it's better to avoid unrelated changes.

Comment (petrhosek) [reply libcxx/src/support/runtime/exception_pointer_msvc.ipp]: Done.

**Comment by StephanTLavavej** [comment]
Awesome, thank you so much for landing this! Please let me know if we can be of any other assistance, or if you find bugs in our now-shared implementation :joy_cat:

File: libcxx/docs/ReleaseNotes/22.rst [hunk 8]
```diff
@@ -128,6 +128,9 @@ Potentially breaking changes
   ``_LIBCPP_ABI_NO_ITERATOR_BASES``. If you are using this flag and care about ABI stability, you should set
   ``_LIBCPP_ABI_NO_REVERSE_ITERATOR_SECOND_MEMBER`` as well.
 
+- ``std::exception_ptr`` on Windows no longer relies on STL and is now implemented entirely within libc++.
```
Comment (mstorsjo) [inline libcxx/docs/ReleaseNotes/22.rst:131]: This release note was added in the file for release 22, while the 22 release already is done. I presume this should be moved to the file for release 23?

Secondly, I'd request to clarify the terminology here. The MSVC C++ ABI is not the only ABI we support on Windows. In mingw mode, we never relied on the MS STL.

So when moving the release note to the 23 file, can you also reword this to say "in MSVC mode" instead of "on Windows", or something along those lines?

**Comment by mstorsjo** [comment]
After having this merged now, I presume we should stop linking against msvcprt/libcpmt? We have references to this in libcxx/CMakeLists.txt and in test/configs/llvm-libc++-*-clangcl.cfg.in.

File: libcxx/docs/ReleaseNotes/22.rst (hunk 8 above)
Comment (petrhosek) [reply libcxx/docs/ReleaseNotes/22.rst:131]: Thanks for catching this, I created https://github.com/llvm/llvm-project/pull/208883 to move the notes.

**Comment by petrhosek** [comment]
> After having this merged now, I presume we should stop linking against msvcprt/libcpmt? We have references to this in libcxx/CMakeLists.txt and in test/configs/llvm-libc++-*-clangcl.cfg.in.

Not yet, we also need to merge https://github.com/llvm/llvm-project/pull/150182.

**Comment by nico** [comment]
We're seeing

```
In file included from ../../third_party/libc++/src/src/exception.cpp:16:
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(149,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  149 |     __call_member_function_2(__dest, __copy_func, __adjusted, 1);
      |                                      ^~~~~~~~~~~
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(151,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  151 |     __call_member_function_1(__dest, __copy_func, __adjusted);
      |                                      ^~~~~~~~~~~
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(207,77): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  207 |           __call_member_function_0(__cpp_eh_record.params.pExceptionObject, __throw_info->pmfnUnwind);
      |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

with this.

**Comment by petrhosek** [comment]
> We're seeing
> 
> ```
> In file included from ../../third_party/libc++/src/src/exception.cpp:16:
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(149,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   149 |     __call_member_function_2(__dest, __copy_func, __adjusted, 1);
>       |                                      ^~~~~~~~~~~
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(151,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   151 |     __call_member_function_1(__dest, __copy_func, __adjusted);
>       |                                      ^~~~~~~~~~~
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(207,77): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   207 |           __call_member_function_0(__cpp_eh_record.params.pExceptionObject, __throw_info->pmfnUnwind);
>       |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~
> 3 errors generated.
> ```
> 
> with this.

Do you have link to full log or instructions for reproducing this? What compiler are you using?

**Comment by nico** [comment]
log: https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8676372941358177969/+/u/compile__with_patch_/stdout?format=raw

(look for "failed:")

It uses clang-cl.

Failing build command:

```
[8311/117871] 28.00s E CXX obj/buildtools/third_party/libc++/libc++/exception.obj	
FAILED: 8bf031e0-eb2f-4d4d-8441-0f3c69035378 "./obj/buildtools/third_party/libc++/libc++/exception.obj" CXX obj/buildtools/third_party/libc++/libc++/exception.obj
err: exit=1
..\..\third_party\llvm-build\Release+Asserts\bin\clang-cl.exe /c ../../third_party/libc++/src/src/exception.cpp /Foobj/buildtools/third_party/libc++/libc++/exception.obj /nologo /showIncludes:user /winsysroot../../third_party/depot_tools/win_toolchain/vs_files/e66617bc68 "-DCR_CLANG_REVISION=\"llvmorg-23-init-19482-g53d18800-1\"" -DCOMPONENT_BUILD -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE -D_LIBCPP_INSTRUMENTED_WITH_ASAN=0 -DCR_LIBCXX_REVISION=9d57af17811b286bda053a9887db849e6ab21482 -D__STD_C -D_CRT_RAND_S -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D_ATL_NO_OPENGL -D_WINDOWS -DCERT_CHAIN_PARA_HAS_EXTRA_FIELDS -DPSAPI_VERSION=2 -DWIN32 -D_SECURE_ATL -DWINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP -DDCHECK_ALWAYS_ON=1 -DUSE_AURA=1 -DWIN32_LEAN_AND_MEAN -DNOMINMAX -D_UNICODE -DUNICODE -DWINCRYPT_USE_SYMBOL_PREFIX -D_DEBUG -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNTDDI_VERSION=NTDDI_WIN7 -D_WIN32_WINNT=_WIN32_WINNT_WIN7 -DWINVER=_WIN32_WINNT_WIN7 -D_LIBCPP_BUILDING_LIBRARY -D_CRT_NONSTDC_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS -DLIBC_NAMESPACE=__llvm_libc_cr -I../../third_party/libc++/src/src -I../.. -Igen -I../../buildtools/third_party/libc++ -I../../third_party/llvm-libc/src -fno-delete-null-pointer-checks /clang:-fno-strict-overflow -fno-ident /clang:-fno-math-errno -fcolor-diagnostics /clang:-fdiagnostics-show-inlining-chain -fmerge-all-constants -fcrash-diagnostics-dir=../clang-crashreports -fno-lifetime-dse -mllvm -instcombine-lower-dbg-declare=0 -mllvm -split-threshold-for-reg-with-hint=0 /clang:-ffp-contract=off -fcomplete-member-pointers /Gy /FS /bigobj /utf-8 /Zc:twoPhase -ffile-reproducible /D__WRL_ENABLE_FUNCTION_STATICS__ -fmsc-version=1934 /Zc:dllexportInlines- -m32 -msse3 /Brepro -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -ffile-compilation-dir=. -no-canonical-prefixes --warning-suppression-mappings=../../build/config/warning_suppression.txt -ftrivial-auto-var-init=pattern /Od /Ob0 /GF -fsanitize=array-bounds -fsanitize-trap=array-bounds -fsanitize-ignore-for-ubsan-feature=array-bounds -Wno-unsafe-buffer-usage-in-static-sized-array -fsanitize=return -fsanitize-trap=return -fsanitize-ignore-for-ubsan-feature=return /guard:cf,nochecks /MDd -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang check-stack-allocated -Xclang -plugin-arg-find-bad-constructs -Xclang relax-ctor-checks-for-aggregates -Xclang -add-plugin -Xclang raw-ptr-plugin -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ptr-to-stack-allocated -Xclang -plugin-arg-raw-ptr-plugin -Xclang disable-check-raw-ptr-to-stack-allocated-error -Xclang -plugin-arg-raw-ptr-plugin -Xclang raw-ptr-exclude-path=../../third_party/dawn/ -Xclang -plugin-arg-raw-ptr-plugin -Xclang raw-ptr-exclude-path=../../build/config/clang/chrome/test/fuzzing/webidl_fuzzing/ -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ptr-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-span-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ref-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang no-special-treatment-for-oilpanized-paths -DUNSAFE_BUFFERS_BUILD -Wunsafe-buffer-usage -Wno-error=unsafe-buffer-usage -Xclang -add-plugin -Xclang unsafe-buffers -Xclang -plugin-arg-unsafe-buffers -Xclang ../../build/config/unsafe_buffers_paths.txt -fstrict-aliasing -Wundef -Wno-pragma-clang-attribute -Wno-ignored-attributes -Wno-macro-redefined /W3 /wd4800 /wd4267 -Wno-unused-variable -Wno-unused-template -Wno-c++11-narrowing -Wno-unused-but-set-variable -Wno-missing-field-initializers -Wno-unused-parameter -Wno-psabi -Wloop-analysis -Wno-unneeded-internal-declaration -Wno-cast-function-type -Wno-thread-safety-reference-return -Wno-nontrivial-memcall -Wno-uninitialized-const-pointer -Wno-unused-but-set-global -Wno-nonportable-include-path -Wexit-time-destructors /WX -Wno-exit-time-destructors -Wno-exit-time-destructors /std:c++23preview -Wno-trigraphs -I../../third_party/libc++/src/include -std:c++latest -Wno-invalid-offsetof -Wenum-compare-conditional -Wno-nullability-completeness -Wunique-object-duplication /EHsc /GR /TP  "-fmodule-name=//buildtools/third_party/libc++:libc++_Private" /Fd"obj/buildtools/third_party/libc++/libc++_cc.pdb"
build step: cxx "./obj/buildtools/third_party/libc++/libc++/exception.obj"
siso_rule: clang-cl/cxx
stderr:
In file included from ../../third_party/libc++/src/src/exception.cpp:16:
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(149,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  149 |     __call_member_function_2(__dest, __copy_func, __adjusted, 1);
      |                                      ^~~~~~~~~~~
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(151,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  151 |     __call_member_function_1(__dest, __copy_func, __adjusted);
      |                                      ^~~~~~~~~~~
../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(207,77): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
  207 |           __call_member_function_0(__cpp_eh_record.params.pExceptionObject, __throw_info->pmfnUnwind);
      |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

**Comment by petrhosek** [comment]
> log: https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8676372941358177969/+/u/compile__with_patch_/stdout?format=raw
> 
> (look for "failed:")
> 
> It uses clang-cl.
> 
> Failing build command:
> 
> ```
> [8311/117871] 28.00s E CXX obj/buildtools/third_party/libc++/libc++/exception.obj	
> FAILED: 8bf031e0-eb2f-4d4d-8441-0f3c69035378 "./obj/buildtools/third_party/libc++/libc++/exception.obj" CXX obj/buildtools/third_party/libc++/libc++/exception.obj
> err: exit=1
> ..\..\third_party\llvm-build\Release+Asserts\bin\clang-cl.exe /c ../../third_party/libc++/src/src/exception.cpp /Foobj/buildtools/third_party/libc++/libc++/exception.obj /nologo /showIncludes:user /winsysroot../../third_party/depot_tools/win_toolchain/vs_files/e66617bc68 "-DCR_CLANG_REVISION=\"llvmorg-23-init-19482-g53d18800-1\"" -DCOMPONENT_BUILD -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE -D_LIBCPP_INSTRUMENTED_WITH_ASAN=0 -DCR_LIBCXX_REVISION=9d57af17811b286bda053a9887db849e6ab21482 -D__STD_C -D_CRT_RAND_S -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D_ATL_NO_OPENGL -D_WINDOWS -DCERT_CHAIN_PARA_HAS_EXTRA_FIELDS -DPSAPI_VERSION=2 -DWIN32 -D_SECURE_ATL -DWINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP -DDCHECK_ALWAYS_ON=1 -DUSE_AURA=1 -DWIN32_LEAN_AND_MEAN -DNOMINMAX -D_UNICODE -DUNICODE -DWINCRYPT_USE_SYMBOL_PREFIX -D_DEBUG -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNTDDI_VERSION=NTDDI_WIN7 -D_WIN32_WINNT=_WIN32_WINNT_WIN7 -DWINVER=_WIN32_WINNT_WIN7 -D_LIBCPP_BUILDING_LIBRARY -D_CRT_NONSTDC_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS -DLIBC_NAMESPACE=__llvm_libc_cr -I../../third_party/libc++/src/src -I../.. -Igen -I../../buildtools/third_party/libc++ -I../../third_party/llvm-libc/src -fno-delete-null-pointer-checks /clang:-fno-strict-overflow -fno-ident /clang:-fno-math-errno -fcolor-diagnostics /clang:-fdiagnostics-show-inlining-chain -fmerge-all-constants -fcrash-diagnostics-dir=../clang-crashreports -fno-lifetime-dse -mllvm -instcombine-lower-dbg-declare=0 -mllvm -split-threshold-for-reg-with-hint=0 /clang:-ffp-contract=off -fcomplete-member-pointers /Gy /FS /bigobj /utf-8 /Zc:twoPhase -ffile-reproducible /D__WRL_ENABLE_FUNCTION_STATICS__ -fmsc-version=1934 /Zc:dllexportInlines- -m32 -msse3 /Brepro -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -ffile-compilation-dir=. -no-canonical-prefixes --warning-suppression-mappings=../../build/config/warning_suppression.txt -ftrivial-auto-var-init=pattern /Od /Ob0 /GF -fsanitize=array-bounds -fsanitize-trap=array-bounds -fsanitize-ignore-for-ubsan-feature=array-bounds -Wno-unsafe-buffer-usage-in-static-sized-array -fsanitize=return -fsanitize-trap=return -fsanitize-ignore-for-ubsan-feature=return /guard:cf,nochecks /MDd -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang check-stack-allocated -Xclang -plugin-arg-find-bad-constructs -Xclang relax-ctor-checks-for-aggregates -Xclang -add-plugin -Xclang raw-ptr-plugin -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ptr-to-stack-allocated -Xclang -plugin-arg-raw-ptr-plugin -Xclang disable-check-raw-ptr-to-stack-allocated-error -Xclang -plugin-arg-raw-ptr-plugin -Xclang raw-ptr-exclude-path=../../third_party/dawn/ -Xclang -plugin-arg-raw-ptr-plugin -Xclang raw-ptr-exclude-path=../../build/config/clang/chrome/test/fuzzing/webidl_fuzzing/ -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ptr-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-span-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang check-raw-ref-fields -Xclang -plugin-arg-raw-ptr-plugin -Xclang no-special-treatment-for-oilpanized-paths -DUNSAFE_BUFFERS_BUILD -Wunsafe-buffer-usage -Wno-error=unsafe-buffer-usage -Xclang -add-plugin -Xclang unsafe-buffers -Xclang -plugin-arg-unsafe-buffers -Xclang ../../build/config/unsafe_buffers_paths.txt -fstrict-aliasing -Wundef -Wno-pragma-clang-attribute -Wno-ignored-attributes -Wno-macro-redefined /W3 /wd4800 /wd4267 -Wno-unused-variable -Wno-unused-template -Wno-c++11-narrowing -Wno-unused-but-set-variable -Wno-missing-field-initializers -Wno-unused-parameter -Wno-psabi -Wloop-analysis -Wno-unneeded-internal-declaration -Wno-cast-function-type -Wno-thread-safety-reference-return -Wno-nontrivial-memcall -Wno-uninitialized-const-pointer -Wno-unused-but-set-global -Wno-nonportable-include-path -Wexit-time-destructors /WX -Wno-exit-time-destructors -Wno-exit-time-destructors /std:c++23preview -Wno-trigraphs -I../../third_party/libc++/src/include -std:c++latest -Wno-invalid-offsetof -Wenum-compare-conditional -Wno-nullability-completeness -Wunique-object-duplication /EHsc /GR /TP  "-fmodule-name=//buildtools/third_party/libc++:libc++_Private" /Fd"obj/buildtools/third_party/libc++/libc++_cc.pdb"
> build step: cxx "./obj/buildtools/third_party/libc++/libc++/exception.obj"
> siso_rule: clang-cl/cxx
> stderr:
> In file included from ../../third_party/libc++/src/src/exception.cpp:16:
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(149,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   149 |     __call_member_function_2(__dest, __copy_func, __adjusted, 1);
>       |                                      ^~~~~~~~~~~
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(151,38): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   151 |     __call_member_function_1(__dest, __copy_func, __adjusted);
>       |                                      ^~~~~~~~~~~
> ../../third_party/libc++/src/src\support/runtime/exception_pointer_msvc.ipp(207,77): error: implicit conversion between pointer-to-function and pointer-to-object is a Microsoft extension [-Werror,-Wmicrosoft-cast]
>   207 |           __call_member_function_0(__cpp_eh_record.params.pExceptionObject, __throw_info->pmfnUnwind);
>       |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~
> 3 errors generated.
> ```

https://github.com/llvm/llvm-project/pull/210570
